Replace GPL-licensed Readline with BSD-licensed Editline

andrewvaughanj opened the pull request:
Initial work to address #4571.

Some points:

- [x] Do we want to preserve `readline` or should we remove it all together?

- [x] Do we need a `contrib/get-libedit` script?

- [x] This is currently 'hard-coded' to assume that `READLINE_COMPENTRY_FUNC_RETURNS_CHARP` always applies when using `libedit`

- [ ] Check that this works on Windows (does CVC4 support `readline` on Windows?)

- [x] Discuss the changes made to this branch by `clang-format`

- [x] Any way to test this? It would probably be okay to create a Python script that sent some input, sent up, and then confirmed we got the same characters

- [x] Do we want to add this configuration to the GH CI?

4tXJ7f commented:
Thanks a lot for doing this!

>  Do we want to preserve readline or should we remove it all together?

I'd be in favor of removing readline if libedit works well enough.

> Do we need a contrib/get-libedit script?

We generally only add a contrib script if there is a good reason, mainly in cases where a dependency is not widely available on package repositories (e.g. ANTLR 3 because it is too old).

> This is currently 'hard-coded' to assume that READLINE_COMPENTRY_FUNC_RETURNS_CHARP always applies when using libedit

Do you think that this is a safe assumption? I'll take a closer look.

> Check that this works on Windows (does CVC4 support readline on Windows?)

I can take care of that and get back to you.

>  Discuss the changes made to this branch by clang-format

I took a cursory look at the formatting changes and they look ok to me. Is there anything in particular that you think could be controversial?

>  Any way to test this? It would probably be okay to create a Python script that sent some input, sent up, and then confirmed we got the same characters

As far as I know, we currently don't test readline in any way but your idea with the Python script sounds good to me.

> Do we want to add this configuration to the GH CI?

I don't think that we want to add a separate configuration but maybe we can add libedit to one of the existing configurations, especially if we have a way to test it. If it makes our CI much slower (I'd be surprised if it did), then we could test it as part of our nightly tests. We will probably also want to change all our nightly builds to include libedit.

andrewvaughanj commented:
> > Do we want to preserve readline or should we remove it all together?
> 
> I'd be in favor of removing readline if libedit works well enough.
> 

I'm happy to rip-out all of the readline code; do you want me to do this in this PR?


> > Do we need a contrib/get-libedit script?
> 
> We generally only add a contrib script if there is a good reason, mainly in cases where a dependency is not widely available on package repositories (e.g. ANTLR 3 because it is too old).
> 

Okay, nothing to do here then!


> > This is currently 'hard-coded' to assume that READLINE_COMPENTRY_FUNC_RETURNS_CHARP always applies when using libedit
> 
> Do you think that this is a safe assumption? I'll take a closer look.
> 

Well, it seems that the `READLINE_COMPENTRY_FUNC_RETURNS_CHARP` stuff is something we explicitly detect (using a CMake test) -- however, this seems like the default for `libedit`.

My guess is here is that `readline` now defaults to returning a `char*`, but when `readline` was originally integrated in CVC4, it used the older `int`-based return.

If we care, I can add a similar test to the `libedit` CMake script, I just don't know if it is neccessary.


> > Check that this works on Windows (does CVC4 support readline on Windows?)
> 
> I can take care of that and get back to you.
> 

I looked into this, and it seems that `readline` (the GNU version) doesn't really work well on Windows, so I presume ths is a non-issue.


> > Discuss the changes made to this branch by clang-format
> 
> I took a cursory look at the formatting changes and they look ok to me. Is there anything in particular that you think could be controversial?
> 

I was just surprised that:

```
} else {
```

got turned into:

```
}
else
{
```

that was all.

Also, what's the style for comments after `endif`s?


> > Any way to test this? It would probably be okay to create a Python script that sent some input, sent up, and then confirmed we got the same characters
> 
> As far as I know, we currently don't test readline in any way but your idea with the Python script sounds good to me.
> 

I can take a look at this.


> > Do we want to add this configuration to the GH CI?
> 
> I don't think that we want to add a separate configuration but maybe we can add libedit to one of the existing configurations, especially if we have a way to test it. If it makes our CI much slower (I'd be surprised if it did), then we could test it as part of our nightly tests. We will probably also want to change all our nightly builds to include libedit.

I think it would be nice to add `libedit-devel` + `--libedit` to one configuration, at least to check that CVC4 with `libedit` compiles/links.

andrewvaughanj commented:
@4tXJ7f I gave a shot at writing a Python script to interact with a version of CVC4 built with line editing. Take a look:

- [`interactive_shell.py`](https://github.com/andrewvaughanj/CVC4/blob/libedit/test/unit/main/interactive_shell.py)

This requires the Python module [`Pexpect`](https://pexpect.readthedocs.io/en/stable/), and it isn't "hooked in" at all to the build/tests.

I've tested this against a non-readline CVC4 (i.e., `master`) and a libedit CVC4 (i.e., this branch) and it _fails_ on `master` and works on `libedit`!

aniemetz commented:
@andrewvaughanj formatting is as intended, that's what it should be.

4tXJ7f commented:
> > > Do we want to preserve readline or should we remove it all together?
> > 
> > 
> > I'd be in favor of removing readline if libedit works well enough.
> 
> I'm happy to rip-out all of the readline code; do you want me to do this in this PR?

Yes, I think that that'd be fine.

> > > This is currently 'hard-coded' to assume that READLINE_COMPENTRY_FUNC_RETURNS_CHARP always applies when using libedit
> > 
> > 
> > Do you think that this is a safe assumption? I'll take a closer look.
> 
> Well, it seems that the `READLINE_COMPENTRY_FUNC_RETURNS_CHARP` stuff is something we explicitly detect (using a CMake test) -- however, this seems like the default for `libedit`.
> 
> My guess is here is that `readline` now defaults to returning a `char*`, but when `readline` was originally integrated in CVC4, it used the older `int`-based return.
> 
> If we care, I can add a similar test to the `libedit` CMake script, I just don't know if it is neccessary.

I briefly looked into this and I think that it is fine to omit it for now.

> > > Check that this works on Windows (does CVC4 support readline on Windows?)
> > 
> > 
> > I can take care of that and get back to you.
> 
> I looked into this, and it seems that `readline` (the GNU version) doesn't really work well on Windows, so I presume ths is a non-issue.

I agree, it is fine to add support for libedit on Windows later if the need arises.

> > > Do we want to add this configuration to the GH CI?
> > 
> > 
> > I don't think that we want to add a separate configuration but maybe we can add libedit to one of the existing configurations, especially if we have a way to test it. If it makes our CI much slower (I'd be surprised if it did), then we could test it as part of our nightly tests. We will probably also want to change all our nightly builds to include libedit.
> 
> I think it would be nice to add `libedit-devel` + `--libedit` to one configuration, at least to check that CVC4 with `libedit` compiles/links.

Yes, I agree.

> I gave a shot at writing a Python script to interact with a version of CVC4 built with line editing.

Looks great. I'll bring up the question of how to best integrate it with our infrastructure during our next Tuesday meeting.

andrewvaughanj commented:
Okay, so I think most of the "important" bits have been done ... I've now ripped-out readline (and `READLINE_COMPENTRY_FUNC_RETURNS_CHARP`) completely.

I've updated any documentation that talks to about readline to either have those bits removed, or have them updated to talk about libedit.

The thing I would like some guidance on, if possible, is how to integrate this into the CI system. I've written the Python script, but I am not sure where it should "live" or how it should be integrated in the "most suitable way".

4tXJ7f commented on test/unit/main/interactive_shell.py:
```diff
@@ -0,0 +1,64 @@
+## \brief A simpel test file to interact with CVC4 with line editing
+#####################
+
+import pexpect
```
Could we make this import conditional? If the user has the dependency, we run the test, otherwise we skip it.

andrewvaughanj replied:
Done!

4tXJ7f commented on test/unit/main/interactive_shell.py:
```diff
@@ -0,0 +1,64 @@
+#!/usr/bin/env python3
```
We've discussed it offline and the best place for this test would be in `test/system`. Could you move it there and modify the corresponding CMakeLists.txt file to add the test? It should be something like:

```
add_test(interactive_shell interactive_shell.py)
```

andrewvaughanj replied:
Done!

4tXJ7f commented on CMakeLists.txt:
```diff
@@ -272,7 +272,6 @@ if(ENABLE_BEST)
   cvc4_set_option(USE_CLN ON)
   cvc4_set_option(USE_CRYPTOMINISAT ON)
   cvc4_set_option(USE_GLPK ON)
```
Let's use `USE_EDITLINE` in out "best" configuration, too.

andrewvaughanj replied:
Done!

4tXJ7f commented on INSTALL.md:
```diff
@@ -175,18 +175,12 @@ and-inverter-graphs (AIG) and ABC is used to simplify these AIGs.
-If you choose to use CVC4 with GNU Readline support, you are licensing CVC4
-under that same license.
-(Usually CVC4's license is more permissive; see above discussion.)
+The [Editline Library ](https://thrysoee.dk/editline/) library is optionally
```
```suggestion
The [Editline Library](https://thrysoee.dk/editline/) library is optionally
```

andrewvaughanj replied:
Done!

4tXJ7f commented on cvc4autoconfig.h.in:
```diff
@@ -43,8 +43,11 @@
+#cmakedefine01 HAVE_LIBEDITLINE
+
+/* Define to 1 to use line editing (in general). */
+#cmakedefine01 HAVE_LINEEDITING
```
Not sure if we need to make this distinction since we only have support for libedit now. I have a slight preference for only `HAVE_LIBEDITLINE` but I can live with two defines as well.

andrewvaughanj replied:
Done!

andrewvaughanj commented:
If you want to check the test: `ctest -VV -R "interactive_shell"`

4tXJ7f commented on test/system/CMakeLists.txt:
```diff
@@ -33,4 +33,16 @@ cvc4_add_system_test(sep_log_api)
 cvc4_add_system_test(statistics)
 cvc4_add_system_test(two_smt_engines)
+
+find_package(PythonInterp)
+if(PYTHONINTERP_FOUND)
+  # Add our interactive shell test _only_ if we have Python
+  add_test(
+    NAME interactive_shell
+    COMMAND
+    "${PYTHON_EXECUTABLE}" "${CMAKE_SOURCE_DIR}/test/system/interactive_shell.py"
+    WORKING_DIRECTORY "${CMAKE_BINARY_DIR}"
+  )
+endif()
```
`${PYTHON_EXECUTABLE}` should already be [defined](https://github.com/CVC4/CVC4/blob/master/CMakeLists.txt#L348-L354) (and we require Python to build CVC4 anyway), so I think that it would be sufficient to just do this:

```suggestion
add_test(
  NAME interactive_shell
  COMMAND
  "${PYTHON_EXECUTABLE}" "${CMAKE_SOURCE_DIR}/test/system/interactive_shell.py"
  WORKING_DIRECTORY "${CMAKE_BINARY_DIR}"
)
```

andrewvaughanj replied:
I did it that way because of this: https://github.com/CVC4/CVC4/blob/master/cmake/FindCython.cmake#L38

andrewvaughanj replied:
Done!

4tXJ7f commented on test/system/interactive_shell.py:
```diff
@@ -0,0 +1,96 @@
+
+    # If we have pexpect ...
+    if "pexpect" in sys.modules:
+        # presume the test failed (e.g., due to an exception)
+        ret_val = 1
```
I think that if there is an exception, we will not actually proceed to `sys.exit(ret_val)` or am I missing something? If I understood it correctly, then setting `ret_val = 1` here is essentially a noop and could be removed.

andrewvaughanj replied:
My original code looked like this:

```
if "pexpect" in sys.modules:
     sys.exit(check_iteractive_shell())

sys.exit(0)
```

shall I just go back to that?

andrewvaughanj replied:
Done!

andrewvaughanj commented:
> Thanks a lot for the updates and your patience! Just two more minor comments, I think that we can merge this PR after that.

What about adding a `libedit` configuration to CI?

andrewvaughanj commented:
I think I would prefer to use CMake to detect `pexpect` rather than doing it at run-time in the Python.

Something like this: https://github.com/llvm/llvm-project/blob/f57dd41562340b0d5d9931b3f1becca761a54d26/llvm/cmake/config-ix.cmake#L634

Thoughts? Happy to do the work if the solution is acceptable.

4tXJ7f commented:
> I think I would prefer to use CMake to detect pexpect rather than doing it at run-time in the Python.

That sounds like a good idea. We already use a similar mechanism here: https://github.com/CVC4/CVC4/blob/3dfb48b80034a9eb628db641c9cec172e53fa910/src/options/CMakeLists.txt#L2-L13

andrewvaughanj commented:
> > I think I would prefer to use CMake to detect pexpect rather than doing it at run-time in the Python.
> 
> That sounds like a good idea. We already use a similar mechanism here:
> 
> https://github.com/CVC4/CVC4/blob/3dfb48b80034a9eb628db641c9cec172e53fa910/src/options/CMakeLists.txt#L2-L13

Implemented! I also added a guard in the CMake that only runs the interactive test if we have a readline build.

andrewvaughanj commented:
Are you happy if I add this to CI?

I am suggesting that we add `--libedit` to both the `production` and `debug` builds in `workflows/ci.yml`, as well as adding `libedit-dev` + `pexpect` to the dependencies.

Reasonable?

4tXJ7f commented on CMakeLists.txt:
```diff
@@ -136,7 +136,7 @@ cvc4_option(USE_CLN           "Use CLN instead of GMP")
 cvc4_option(USE_GLPK          "Use GLPK simplex solver")
 cvc4_option(USE_KISSAT        "Use Kissat SAT solver")
-cvc4_option(USE_READLINE      "Use readline for better interactive support")
+cvc4_option(USE_EDITLINE      "Use editline for better interactive support")
```
```suggestion
cvc4_option(USE_EDITLINE      "Use Editline for better interactive support")
```

andrewvaughanj replied:
Done!

4tXJ7f commented:
One more minor request: Could you please add a sentence to the NEWS file about this change? We try to be more proactive about updating the NEWS file as it makes releases easier.

andrewvaughanj commented:
> Sounds great! Thanks a lot for your patience with this.

Done!

Now it is time for 🤞 


> One more minor request: Could you please add a sentence to the NEWS file about this change? We try to be more proactive about updating the NEWS file as it makes releases easier.

Done!

andrewvaughanj commented:
Okay, so to revisit a _much earlier_ point, we *do* need `READLINE_COMPENTRY_FUNC_RETURNS_CHARP` for macOS:

```
[ 97%] Building CXX object src/main/CMakeFiles/main.dir/interactive_shell.cpp.o
4831
/Users/runner/work/CVC4/CVC4/src/main/interactive_shell.cpp:103:38: error: assigning to 'Function *' (aka 'int (*)(const char *, int)') from incompatible type 'char *(const char *, int)': different return type ('int' vs 'char *')
4832
    ::rl_completion_entry_function = commandGenerator;
4833
                                     ^~~~~~~~~~~~~~~~
4834
1 error generated.
```

I'll try to fix this today.

andrewvaughanj commented:
Hmm, so a libedit installed from MacPorts _does not_ have this issue; looks like maybe it is only the system one that needs this check. Will carry on looking.

andrewvaughanj commented:
As a side note, it seemed that the logic for `READLINE_COMPENTRY_FUNC_RETURNS_CHARP` didn't really work if there were multiple versions of `readline` installed.

The CMake test set the library directory but not the include paths; this meant that the test compiled with the "system" header but linked against the "user" library (e.g., on macOS this would be the library from `brew` or MacPorts). What you'd get is that the test would pass (and CMake would say "does not have a `char*` return") but then the CVC4 build would fail when it used the `rl_completion_entry_function` with an `int` return.

This is now resolved in this branch.

andrewvaughanj commented:
`interactive_shell` ran and passed on: ubuntu-latest:production, ubuntu-latest:debug, macos-latest:production and macos-latest:debug 💪

andrewvaughanj commented:
@4tXJ7f how's this looking (from your perspective)? Is there anything outstanding you need to me to look at?

andrewvaughanj commented:
Sure! **However** isn't CVC4's PR closing mechanism on GH to squash+rebase anyway? That's why I've stopped squashing the commits ...

andrewvaughanj commented:
Also, what's with the change to COPYING? Did was that introduced in a merge? That *is not* in my `HEAD` (which is behind CVC4's `HEAD`).

andrewvaughanj commented:
Okay, I fixed those changes by doing the rebase (I can see how/why they were introduced).

I'm still not sure about the need to manually squash though (it is 100% fine for me to do so, I just don't know if it is necessary).

4tXJ7f commented:
> Okay, I fixed those changes by doing the rebase (I can see how/why they were introduced).

Sorry, that was me. Thanks for fixing.

> I'm still not sure about the need to manually squash though (it is 100% fine for me to do so, I just don't know if it is necessary).

Yeah, me neither and a quick Google search didn't really yield anything. I brought it up for discussion on our Slack channel. It would make sense to me that I can integrate your changes into the code base since you signed off on all of them. What feels a bit iffy is the fact that I would sign the squashed commit in your name.

andrewvaughanj commented:
> Yeah, me neither and a quick Google search didn't really yield anything. I brought it up for discussion on our Slack channel. It would make sense to me that I can integrate your changes into the code base since you signed off on all of them. What feels a bit iffy is the fact that I would sign the squashed commit in your name.

💥 all squashed 💥

4tXJ7f commented:
It's in! Finally :)
